packages/medusa/src/interfaces/services/event-bus.ts

```diff
@@ -1,6 +1,120 @@
 import { EntityManager } from "typeorm"
+import { ulid } from "ulid"
+import { StagedJob } from "../../models"
+
+export type Subscriber<T = unknown> = (
+  data: T,
+  eventName: string
+) => Promise<void>
+
+export type SubscriberContext = {
+  subscriberId: string
+}
+
+export type SubscriberDescriptor = {
+  id: string
+  subscriber: Subscriber
+}
+
+export type EventHandler<T = unknown> = (
+  data: T,
+  eventName: string
+) => Promise<void>
 
 export interface IEventBusService {
-  emit(event: string, data: any): Promise<void>
-  withTransaction(transactionManager?: EntityManager): this
+  emit<T>(
+    eventName: string,
+    data: T,
+    options?: unknown
+  ): Promise<StagedJob | void>
+
+  withTransaction(manager: EntityManager): this
+}
+
+export interface IEventBusModuleService {
+  emit<T>(
+    eventName: string,
+    data: T,
+    options?: unknown
+  ): Promise<StagedJob | void>
+
+  subscribe(
+    event: string | symbol,
+    subscriber: Subscriber,
+    context?: SubscriberContext
+  ): this
+
+  unsubscribe(
+    event: string | symbol,
+    subscriber: Subscriber,
+    context?: SubscriberContext
+  ): this
+}
+
+export abstract class AbstractEventBusModuleService
+  implements IEventBusModuleService
+{
+  public eventToSubscribersMap_: Map<string | symbol, SubscriberDescriptor[]> =
```

Comment:
todo: protected

Reply:
and maybe add `get eventToSubscribersMap`

Reply:
we don't want it to be re assigned externally

Reply:
Like so?
```
protected get eventToSubscribersMap(): Map<
    string | symbol,
    SubscriberDescriptor[]
  > {
    return this.eventToSubscribersMap_
  }
```

Reply:
No it was more of making the prop protected and having a getter public. Because here the prop is public and I assume that it was only to access it in the tests.